### DISCUSSION TITLE
Proper inter-process sync (e.g. named mutex) between C++ & C# .NET 7 processes on Win & MacOS

### DISCUSSION THREAD

**Discussion by Volodymyr-Kashyn** [dotnet/runtime#87475] [Q&A]
Hi!

Is there any simple cross-platform Win&MacOS way of syncing two processes: Native C++ and .NET 7 app (e.g. MAUI)?

Problem: synchronize read-write access to a file. Both processes can read & write.

For Windows it is pretty easy. Named mutex is used:
first, initialize [System.Threading.Mutex](https://learn.microsoft.com/en-us/dotnet/api/system.threading.mutex?view=net-7.0) with a name arg in .NET app and use [CreateMutexEx](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-createmutexexa) in native C++ app.

However, I cannot find proper easy solution for MacOS (arm & Intel). 
What I understand is that named version of [System.Threading.Mutex](https://learn.microsoft.com/en-us/dotnet/api/system.threading.mutex?view=net-7.0) is based on file locks in MacOS. However, I cannot find file instance for it, which contains a name of my mutex.
At least, if there's a static file based on my mutex name, probably I can use native `flock` for syncing in C++ app.

Questions:

1. Where can I find locked file which is used for sync in MacOS? 
`/tmp/.dotnet` or `$TMPDIR/.dotnet` does not contain anything similar to my mutex name.

2. Are there any proper, better or even OOB ways for inter-process synchronization between C++ & .NET 7 app for MacOS?

**Comment by MichalPetryka** [comment]
Since the documentation doesn't specify what underlying OS features are used for inter process mutexes, I'd recommend using PInvokes to access the native APIs from C# instead to guarantee it being the same one as the one used in your C++ code.
